src/pages/api/gateway/[gatewayUID]/sensor/[macAddress]/config.ts

```diff
@@ -1,95 +1,100 @@
+    return false;
+  }
+  return true;
+}
+function validateRequest(
   req: NextApiRequest,
   res: NextApiResponse
-) {
-  // Gateway UID must be a string
-  if (typeof req.query.gatewayUID !== "string") {
-    res.status(400).json({ err: HTTP_STATUS.INVALID_GATEWAY });
-    return;
+): false | ValidRequest {
+  const { gatewayUID, macAddress } = req.query;
+  // TODO (carl): figure out what to do about this unsafe assignment of any. I really want runtime
+  // typechecking. All the code below feels terrible to write when we've already specified the API
+  // in typescript. Maybe we should really use a different language to specify the API and generate
+  // the typescript and typechecking code?
+  const { name, location } = req.body;
+
+  if (typeof gatewayUID !== "string") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: HTTP_STATUS.INVALID_GATEWAY });
+    return false;
   }
-  // Sensor MAC must be a string
-  if (typeof req.query.macAddress !== "string") {
-    res.status(400).json({ err: HTTP_STATUS.INVALID_SENSOR_MAC });
-    return;
+  if (typeof macAddress !== "string") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: HTTP_STATUS.INVALID_SENSOR_MAC });
+    return false;
+  }
+  if (typeof location !== "string" && typeof location !== "undefined") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: "location should be a string or undefined" });
+    return false;
+  }
+  if (typeof name !== "string" && typeof name !== "undefined") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: "name should be a string or undefined" });
+    return false;
+  }
+  if (typeof location === "undefined" && typeof name === "undefined") {
+    res.status(StatusCodes.BAD_REQUEST);
+    res.json({ err: HTTP_STATUS.INVALID_SENSOR_CONFIG_BODY });
+    return false;
   }
 
-  // Query params
-  const { gatewayUID, macAddress } = req.query;
-  // Body params
-  const { loc, name } = req.body as NoteSensorConfigBody;
-  // Notehub values
-  const { hubBaseURL, hubAuthToken, hubProjectUID } = config;
-  // API path
-  const endpoint = `${hubBaseURL}/req?project=${hubProjectUID}&device=${gatewayUID}`;
-  // API headers
-  const headers = {
-    [HTTP_HEADER.CONTENT_TYPE]: HTTP_HEADER.CONTENT_TYPE_JSON,
-    [HTTP_HEADER.SESSION_TOKEN]: hubAuthToken,
-  };
+  return { gatewayUID, macAddress, location, name };
+}
 
-  // note.update payload
-  const noteUpdate = {
-    req: "note.update",
-    file: "config.db",
-    note: macAddress,
-    body: {
-      loc,
-      name,
-    },
-  };
+async function performRequest({
+  macAddress,
+  gatewayUID,
+  location,
+  name,
+}: ValidRequest) {
+  const app = services().getAppService();
+  try {
+    if (location) await app.setSensorLocation(gatewayUID, macAddress, location);
+    if (name) await app.setSensorName(gatewayUID, macAddress, name);
+  } catch (cause) {
+    throw new ErrorWithCause("Could not perform request", { cause });
```

Comment:
There’s a `getError()` function in `services` that wraps `ErrorWithCause` to limit our dependency on that to one place.

Reply:
I shouldn't deviate from the rest of the codebase you're right. However, I think it might be better to just use ErrorWithCause since it has the same API as Error will have in the later versions of JS. Thus, it'll be a very easy find+replace when the time comes. It's ok for our code to depend on language features (or anything more stable than we are). Also, I don't love the semantic of `get` when you're actually doing `new` - it just breaks what I expect to see in a catch block.